protocol/smart-contract-versioning.md

```diff
@@ -0,0 +1,86 @@
+# Purpose
+
+This design doc proposes a new smart contracts versioning scheme to improve clarity and more closely align with typical semantic versioning practices.
+
+# Summary
+
+The proposed versioning scheme for contracts will follow a true semver format, with major, minor, and patch versions, as well as beta, release candidate, and feature tags.
+The monorepo releases will also follow semver, with the version bump determined by the highest version bump of any individual contract modified in the release.
+
+# Problem Statement + Context
+
+There are two facets to contract versioning: the versioning of monorepo releases, and the versioning of individual contracts.
+
+- **Monorepo Releases**:  The current versioning scheme of monorepo contract releases does not follow a true semver. For example, the Fault Proofs release introduced breaking changes by changing the ABI, and we are releasing this as a minor version bump from `op-contracts/v1.3.0` to `op-contracts/v1.4.0`. This was not an intentional decision. We likely slid into this path because node releases cannot bump the major version (for go-specific reasons), and our current [contracts release process](https://www.notion.so/oplabs/Minimal-Contracts-Release-Process-ef8e91b1b13944e7a7f632a824895764?pvs=4) does not specify how to determine the version bump for an upcoming monorepo release.
+- **Individual Contracts**: The current versioning scheme of individual contracts is also not a true semver, but in a different way. We do bump major and minor versions appropriately, but we don't distinguish when a contract is in beta, release candidate, or feature status. This makes it difficult to determine the stability and compatibility of a given contract version.
+
+These have resulted in various issues, such as chains deploying in-development versions of contracts.
+
+Unlike typical software development, OP Stack chains cannot rely on "compatible semvers" to compose the chain of various contracts.
+Instead, a given monorepo contracts release must be defined by an exact set of contract versions, not a range of semver-compatible versions.
+
+# Alternatives Considered
+
+1. Keep the current versioning scheme: This alternative would maintain the status quo but would not address the issues of clarity, consistency, and alignment with semver practices.
+
+2. Use semver for contracts but not for monorepo releases: This alternative would improve the versioning of individual contracts but would not provide a clear mapping between monorepo releases and contract versions.
+
+# Proposed Solution
+
+The proposed solution is to more closely adopt a true semver format for both individual contracts and monorepo releases.
+
+We preserve the existing rules for when to bump major, minor, and patch versions in individual contracts:
+
+- `patch` releases are to be used only for changes that do NOT modify contract bytecode (such as updating comments).
```

Comment:
This is so much better without the bindings, just fyi updating the version itself will modify the bytecode so this definition is impossible unless there is a clause where the version itself doesnt count as modifying the bytecode

Reply:
Ah great point. Another idea is change the version from being immutable into a storage variable, that way initcode changes but runtime code does not, and we'd update this to say "that do NOT modify contract runtime bytecode". But I prefer your idea of adding an exception clause, because modifying initcode can still be a significant change

Reply:
Documented this exception in 0f9f896